File: Top_100_Liked_Questions/338._Counting_Bits/solution.js
/**
 * @param {number} num
 * @return {number[]}
 */
var countBits = function(num) {
    // hamming weight method
    let ans = new Array(num+1).fill(0);
    
    for (let i=1; i<num+1; i++) {
        ans[i] = ans[i & (i-1)] + 1;
    }
    
    return ans;
}

var countBits2 = function(num) {
    // array doubling method
    let ans = [0];
    
    while (ans.length < num+1) {
        ans = [...ans, ...ans.map(x => x+1)];
    }
    
    return ans.slice(0,num+1);
};

File: Top_100_Liked_Questions/338._Counting_Bits/solution.java
class Solution {
    public int[] countBits(int num) {
        int[] ans = new int[num + 1];
        for (int i=1; i<=num; i++)
            ans[i] = ans[i & (i - 1)] + 1;

        return ans;
    }

    public int[] countBits2(int num) {
        int[] ans = new int[num+1];

        int upto = 1;
        while (true) {
            for (int i=0; i<upto; i++) {
                if (upto + i == num + 1)
                    return ans;

                ans[upto+i] = ans[i] + 1;
            }

            upto *= 2;
        }
    }
}